source/uwp/Renderer/lib/AdaptiveInputs.cpp

```diff
@@ -29,16 +33,76 @@ namespace AdaptiveNamespace
         return StringToJsonObject(GetInputItemsAsJsonString(), value);
     }
 
-    HRESULT AdaptiveInputs::AddInputValue(_In_ IAdaptiveInputValue* inputValue)
+    HRESULT AdaptiveInputs::AddInputValue(_In_ IAdaptiveInputValue* inputValue, _In_ IAdaptiveRenderArgs* renderArgs)
     {
-        m_inputValues->push_back(inputValue);
+        ComPtr<IAdaptiveRenderArgs> localRenderArgs(renderArgs);
+
+        ComPtr<IAdaptiveCard> parentCard;
+        RETURN_IF_FAILED(localRenderArgs->get_ParentCard(parentCard.GetAddressOf()));
+        ComPtr<AdaptiveCard> parentCardImpl = PeekInnards<AdaptiveNamespace::AdaptiveCard>(parentCard);
+        InternalId cardId = parentCardImpl->GetInternalId();
+
+        ComPtr<IAdaptiveInputElement> inputElement;
+        RETURN_IF_FAILED(inputValue->get_InputElement(inputElement.GetAddressOf()));
+
+        ComPtr<AdaptiveInputElementBase> cardElementImpl;
+        RETURN_IF_FAILED(inputElement->QueryInterface(cardElementImpl.GetAddressOf()));
+
+        InternalId inputId = cardElementImpl->GetInternalId();
+        std::size_t inputIdHash = inputId.Hash();
+
+        m_inputsPerCard[cardId.Hash()].push_back(inputIdHash);
+        m_inputValues[inputIdHash] = inputValue;
+        return S_OK;
+    }
+
+    HRESULT AdaptiveInputs::LinkSubmitActionToCard(_In_ ABI::AdaptiveNamespace::IAdaptiveSubmitAction* submitAction,
+                                                   _In_ ABI::AdaptiveNamespace::IAdaptiveRenderArgs* renderArgs)
+    {
+        ComPtr<IAdaptiveSubmitAction> localSubmitAction(submitAction);
+        ComPtr<AdaptiveNamespace::AdaptiveSubmitAction> submitActionImpl =
+            PeekInnards<AdaptiveNamespace::AdaptiveSubmitAction>(localSubmitAction);
+        InternalId actionId = submitActionImpl->GetInternalId();
+
+        ComPtr<IAdaptiveRenderArgs> localRenderArgs(renderArgs);
+        ComPtr<IAdaptiveCard> adaptiveCard;
+        RETURN_IF_FAILED(localRenderArgs->get_ParentCard(adaptiveCard.GetAddressOf()));
+
+        ComPtr<AdaptiveCard> adaptiveCardImpl = PeekInnards<AdaptiveNamespace::AdaptiveCard>(adaptiveCard);
+        InternalId cardId = adaptiveCardImpl->GetInternalId();
+
+        m_containerCardForAction[actionId.Hash()] = cardId.Hash();
+        return S_OK;
+    }
+
+    HRESULT AdaptiveInputs::LinkCardToParent(_In_ InternalId cardId, _In_ InternalId parentCardId)
+    {
+        m_parentCard[cardId.Hash()] = parentCardId.Hash();
+        return S_OK;
+    }
+
+    HRESULT AdaptiveInputs::GetInputValue(_In_ IAdaptiveInputElement* inputElement, IAdaptiveInputValue** inputValue)
+    {
+        ComPtr<IAdaptiveInputElement> localInputElement(inputElement);
+
+        ComPtr<AdaptiveInputElementBase> inputElementImpl;
+        RETURN_IF_FAILED(localInputElement->QueryInterface(inputElementImpl.GetAddressOf()));
```

Comment:
>localInputElement->QueryInterface [](start = 25, length = 33)

If you're using QI instead of .As you don't need the local com pointer, you can use inputElement directly. Alternatively, most of this code is using the .As pattern, and it may be better to do that here as well for consistency rather than doing it different ways in different places.